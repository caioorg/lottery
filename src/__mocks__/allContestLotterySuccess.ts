export const GET_ALL_CONTEST_LOTTERY_SUCCESS = {
  status: 200,
  data: [
    { loteriaId: 0, concursoId: "2359" },
    { loteriaId: 1, concursoId: "5534" },
    { loteriaId: 2, concursoId: "2200" },
    { loteriaId: 3, concursoId: "2167" },
    { loteriaId: 4, concursoId: "1622" },
    { loteriaId: 5, concursoId: "440" }
  ]
}