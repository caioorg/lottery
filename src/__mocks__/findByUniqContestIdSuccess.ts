export const FIND_BY_UNIQUE_CONTEST_ID_SUCCESS = {
  status: 200,
  data: {
    id:"2359",
    loteria:0,
    numeros:[ "31", "32", "39", "42", "43", "51"],
    data:"2022-03-18T21:58:29.745Z"
  }
}

