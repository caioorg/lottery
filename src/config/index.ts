export default {
  baseUrlRequest: String(
    import.meta.env.VITE_BASE_URL || 
    'https://brainn-api-loterias.herokuapp.com/api/v1'
  )
}