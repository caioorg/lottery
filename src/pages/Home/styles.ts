import styled from "styled-components";
import SideBarSvg from '../../config/assets/sidebar.svg'
import SideBarMobileSvg from '../../config/assets/sidebar-mobile.svg'
import { pixelToRem } from '../../styles/pxToRem'
import { device } from '../../styles/responsive'

interface TypeLottery {
  [key: string]: string
}

const type: TypeLottery = {
  'mega-sena': "var(--mega-sena)",
  quina: 'var(--corner)',
  'lotof√°cil': 'var(--easyLotto)',
  'lotomania': 'var(--lotoMania)',
  'timemania': 'var(--teamMania)',
  'dia-de-sorte': 'var(--luckDay)'
}

export const Container = styled.main`
  display: flex;
  height: 100%;

  @media ${device.mobileS} {
    display: block;
    height: auto;
    width: 100%;
  }
`
interface SectionLotteryProps {
  name?: string
}


export const SectionLottery = styled.section<SectionLotteryProps>`
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  color: #FFF;
  background: ${(props) => props.name ? type[props.name]: '#000'};
  padding: ${pixelToRem(92)};
  mask: url(${SideBarSvg}) no-repeat;
  width: 35%;

  @media ${device.mobileS} {
    width: 100%;
    mask: url(${SideBarMobileSvg}) no-repeat 50%;
    align-items: center;
    padding: ${pixelToRem(15)};
    justify-content: space-evenly;
    min-height: ${pixelToRem(340)};
  }
`

export const TitleLottery = styled.h1`
  font-weight: bold;
  font-size: ${pixelToRem(30)};
  text-transform: uppercase;
`

export const Select = styled.select`
  border-radius: 4px;
  min-width: ${pixelToRem(215.91)};
  min-height: ${pixelToRem(45.2)};
  padding: ${pixelToRem(12)} ${pixelToRem(23)};
  text-transform: uppercase;
  font-weight: 500;
  font-size: ${pixelToRem(15)};
  color: #333333;
  border: 0;
`

export const Contest = styled.p`
  font-weight: 500;
  font-size: ${pixelToRem(14)};
  margin-bottom: ${pixelToRem(14.42)};
  text-transform: uppercase;
`
export const InfoContest = styled.p`
  font-weight: 700;
  font-size: ${pixelToRem(20)}
`

export const SectionLotteryNumbers = styled.section`
  display: flex;
  justify-content: center;
  align-items: center;
  gap: ${pixelToRem(35)};
  position: relative;
  width: 75%;

  @media ${device.mobileS} {
    width: 100%;
    padding: 0 ${pixelToRem(10)};
    flex-direction: column;
    justify-content: space-evenly;
    min-height: ${pixelToRem(360)};
  }
`

export const NumbersLottery = styled.h2`
  font-weight: 700;
  font-size: ${pixelToRem(27)};
  color: #333333;
  width: ${pixelToRem(111.2)};
  height: ${pixelToRem(106.36)};
  border-radius: 50%;
  background: #FFF;
  display: flex;
  justify-content: center;
  align-items: center;

  @media ${device.mobileS} {
    width: ${pixelToRem(66)};
    height: ${pixelToRem(66)};
    font-size: ${pixelToRem(20)};
  }
`

export const ContentNumbers = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  gap: ${pixelToRem(35)};
  max-width: ${pixelToRem(900)};
  flex-wrap: wrap;
  
  @media ${device.mobileS} {
    gap: ${pixelToRem(20)};
    margin: ${pixelToRem(40)} 0 0;
  }
`

export const Instructions = styled.p`
  position: fixed;
  bottom: ${pixelToRem(90)};

  @media ${device.mobileS} {
    position: initial;
    font-size: ${pixelToRem(14)};
    text-align: center;
    margin-bottom: ${pixelToRem(40)};
  }
`

export const ContentTitle = styled.div`
  display: flex;
  align-items: center;
  gap: ${pixelToRem(16)};

  @media ${device.mobileS} {
    flex-direction: column;
  }
`

export const ContestWrapper = styled.div`
  @media ${device.mobileS} {
    display: none;
  }
`
export const ContestWrapperMobile = styled.div`
  display: none;

  @media ${device.mobileS} {
    display: flex;
  }
`