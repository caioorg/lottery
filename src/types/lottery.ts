export interface IResponseAllTypes {
  id: number,
  nome: string
}

export interface IResponseLotteryContest {
  concursoId: string,
  loteriaId: number
}

export interface IResponseFindByIdContest {
  id: string,
  loteria: string | number
  numeros: string[],
  data: string
}

export interface IStateLotteryForm {
  idLottery: string
}

export interface IStateContest {
  id: string | number,
  loteria: number | string,
  numeros: string[],
  data: string
}
