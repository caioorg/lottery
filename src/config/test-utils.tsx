import React from 'react'
import { render, RenderResult, RenderOptions } from '@testing-library/react'
import { GlobalStyles } from '@app/styles/globalStyles'

const ProvidersWithRoute: React.FC = ({ children }) => (
  <>
    <GlobalStyles />
    {children}
  </>
)

export const renderComponent = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: ProvidersWithRoute, ...options })