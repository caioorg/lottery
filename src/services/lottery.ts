import { api } from './api'
import config from '@app/config'
import { Response } from '@app/types/response'
import {
  IResponseAllTypes,
  IResponseLotteryContest,
  IResponseFindByIdContest
} from '@app/types/lottery'

export const getAllTypesLottery = async (): Promise<Response<IResponseAllTypes[]>> => {
  try {
    return await api(config.baseUrlRequest).get('/loterias')
  } catch (error) {
    return Promise.reject(error)
  }
}

export const getAllLotteryContest = async ():
  Promise<Response<IResponseLotteryContest[]>> => {
  try {
    return await api(config.baseUrlRequest).get('/loterias-concursos')
  } catch (error) {
    return Promise.reject(error)
  }
}

export const findByUniqContestId = async (
  id?: string
): Promise<Response<IResponseFindByIdContest>> => {
  try {
    return await api(config.baseUrlRequest).get(`/concursos/${id}`)
  } catch (error) {
    return Promise.reject(error)
  }
}