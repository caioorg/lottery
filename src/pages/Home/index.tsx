import React, { useEffect, useMemo, useState } from 'react';
import { useForm } from 'react-hook-form'
import Branding from '@app/config/assets/logo.svg'
import { getAllLotteryContest, getAllTypesLottery, findByUniqContestId } from '@app/services/lottery'
import { IStateContest, IStateLotteryForm } from '@app/types/lottery'
import {
  SectionLottery,
  Container,
  TitleLottery,
  Select,
  Contest,
  InfoContest,
  SectionLotteryNumbers,
  NumbersLottery,
  ContentNumbers,
  Instructions,
  ContentTitle,
  ContestWrapper,
  ContestWrapperMobile
} from './styles'

const Home: React.FC = () => {
  const [typeLottery, setTypeLottery] = useState<{ id: number, nome: string}[]>([])
  const [lotteryContest, setLotteryContest] = useState<{ loteriaId: number, concursoId: string }[]>([])
  const [contest, setContest] = useState<IStateContest>({ data: '', id: '', loteria: '', numeros: []})
  const { register, watch } = useForm<IStateLotteryForm>({
    defaultValues: { idLottery: '' }
  })

  useEffect(() => {
    async function initialState() {
      try {
        const typesLottery = await getAllTypesLottery()
        const lotteryConstest = await getAllLotteryContest()

        if (typesLottery.status === 200 && typesLottery.data.length > 0) {
          setTypeLottery(typesLottery.data)
        }        

        if (lotteryConstest.status === 200 && lotteryConstest.data.length > 0) {
          setLotteryContest(lotteryConstest.data)
        }
      } catch (error) {
        // console.log(error)
      }
    }

    initialState()
  }, [])

  useEffect(() => {
    const idLottery = watch('idLottery')

    async function findByContestId(id?: string) {
      try {
        const { data, status } = await findByUniqContestId(id)

        if(status === 200 && data) setContest(data)
      } catch (error) {
        // console.log(error)
      }
    }
    
    if (idLottery) {
      const contest = lotteryContest?.find(item => String(item.loteriaId) === idLottery)

      findByContestId(contest?.concursoId)
    }

  }, [watch('idLottery'), lotteryContest])

  const nameLottery = useMemo(() => {
    const name = typeLottery.find(item => String(item.id) === watch('idLottery'))
    return name?.nome || ''
  }, [watch('idLottery')])

  return (
    <Container>
      <SectionLottery name={nameLottery.replaceAll(' ', '-')}>
        <form>
          <Select data-testid='idLottery' {...register('idLottery')}>
            <option value=''></option>
            {typeLottery.map(item => (
              <option key={item.id} value={item.id}>{item.nome}</option>
            ))}
          </Select>  
        </form>
        
        {nameLottery && contest && (
          <>
            <ContentTitle>
              <img src={Branding}  />
              <TitleLottery>{nameLottery}</TitleLottery>
            </ContentTitle>

            <ContestWrapper>
              <Contest>concurso</Contest>
              <InfoContest>
                {contest?.id || ''} - {contest?.data && new Date(contest?.data).toLocaleDateString() || ''}
              </InfoContest>
            </ContestWrapper>

            <ContestWrapperMobile>
              <Contest>Concurso Nº {contest?.id || ''}</Contest>
            </ContestWrapperMobile>
          </>
        )}
        
      </SectionLottery>
      <SectionLotteryNumbers>
        <ContentNumbers>
          {contest.numeros.map((item: string) => <NumbersLottery key={item}>{item}</NumbersLottery>)}
        </ContentNumbers>
        <Instructions>Este sorteio é meramente ilustrativo e não possui nenhuma ligação com a CAIXA.</Instructions>
      </SectionLotteryNumbers>   
    </Container>
  )
}

export default Home;