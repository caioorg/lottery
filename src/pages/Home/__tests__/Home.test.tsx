import React from 'react'
import { waitFor } from '@testing-library/react'
import { renderComponent } from '@app/config/test-utils'
import * as serviceLottery from '@app/services/lottery'
import { GET_ALL_TYPES_LOTTERY_SUCCESS } from '@app/__mocks__/allTypesLotterySuccess'
import { GET_ALL_CONTEST_LOTTERY_SUCCESS } from '@app/__mocks__/allContestLotterySuccess'
import Home from '../index'

const mockGetAllTypesLottery = jest.spyOn(serviceLottery, 'getAllTypesLottery')
mockGetAllTypesLottery.mockResolvedValue(GET_ALL_TYPES_LOTTERY_SUCCESS)

const mockGetAllLotteryConstest = jest.spyOn(serviceLottery, 'getAllLotteryContest')
mockGetAllLotteryConstest.mockResolvedValue(GET_ALL_CONTEST_LOTTERY_SUCCESS)

describe('Pages Home', () => {  

  it('should render properly', async () => {
    await waitFor(() => {
      const { container } =  renderComponent(<Home />)
      expect(container).toMatchSnapshot()
    })
  })
})