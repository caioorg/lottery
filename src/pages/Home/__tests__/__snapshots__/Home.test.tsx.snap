// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages Home should render properly 1`] = `
<div>
  <main
    class="sc-bdvvtL bDGBmc"
  >
    <section
      class="sc-gsDKAQ kkkkeT"
      name=""
    >
      <form>
        <select
          class="sc-hKwDye kzGsAq"
          data-testid="idLottery"
          name="idLottery"
        >
          <option
            value=""
          />
          <option
            value="0"
          >
            mega-sena
          </option>
          <option
            value="1"
          >
            quina
          </option>
          <option
            value="2"
          >
            lotofácil
          </option>
          <option
            value="3"
          >
            lotomania
          </option>
          <option
            value="4"
          >
            timemania
          </option>
          <option
            value="5"
          >
            dia de sorte
          </option>
        </select>
      </form>
      
    </section>
    <section
      class="sc-gKclnd gCSmHZ"
    >
      <div
        class="sc-furwcr jBCBiY"
      />
      <p
        class="sc-pVTFL jrCaeW"
      >
        Este sorteio é meramente ilustrativo e não possui nenhuma ligação com a CAIXA.
      </p>
    </section>
  </main>
</div>
`;
